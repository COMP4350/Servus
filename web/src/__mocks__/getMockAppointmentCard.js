function getMockAppointmentCard() {
    return {
        buyer: 'Risto',
        provider: 'Arveend',
        service_id: '1001',
        created_at: new Date(100),
        booked_time: new Date(500),
    }
}

export const mockAppointmentCard = getMockAppointmentCard();