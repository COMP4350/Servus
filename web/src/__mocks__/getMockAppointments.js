function getMockAppointments() {
    return [
        {
            buyer: 'Risto',
            provider: 'Arveend',
            service_id: '1001',
            created_at: new Date(100),
            booked_time: new Date(500),
        },
        {
            buyer: 'Arveend',
            provider: 'Risto',
            service_id: '1002',
            created_at: new Date(700),
            booked_time: new Date(1200),
        },
        {
            buyer: 'Caden',
            provider: 'Andy',
            service_id: '1003',
            created_at: new Date(1400),
            booked_time: new Date(1900),
        },
    ]
}

function getMockAppointmentsCookie() {
    return {
        username: 'testusername',
    }
}

export const mockAppointments = getMockAppointments();
export const mockAppointmentsCookie = getMockAppointmentsCookie();