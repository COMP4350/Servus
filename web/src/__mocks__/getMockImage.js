function getMockImage() {
    return {
        ownerUsername: 'riki',
        uploadDate: new Date(2012, 1, 1),
        imageName: 'example_image.jpeg',
        imageData: 'the/link/to/data',
        profilePicture: false,
    };
}

export const mockImage = getMockImage();
