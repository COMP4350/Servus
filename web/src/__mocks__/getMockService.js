function getMockService() {
    return {
        _id: '1001',
        name: "Risto's Painting",
        provider: 'Risto',
        description: 'Risto paints you',
        cost: 100,
        duration: '0100',
        availability: [{ weekday: 1, start_time: '0800', end_time: '1800' }],
        ratings: [{ username: 'arvind', rating: 5 }],
    };
}

export const mockService = getMockService();
