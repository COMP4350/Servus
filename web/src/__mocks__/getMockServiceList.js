function getMockServiceList() {
    return [
        {
            name: "Risto's Painting",
            provider: 'Risto',
            description: 'Risto paints you',
            cost: 100,
            duration: 60,
        },
        {
            name: "Reesto's Painting",
            provider: 'Reesto',
            description: 'Reesto paints yee',
            cost: 120,
            duration: 80,
        },
        {
            name: "Rasta's Painting",
            provider: 'Rasta',
            description: 'Rasta paints yaa',
            cost: 150,
            duration: 90,
        },
    ];
}

export const mockServiceList = getMockServiceList();
