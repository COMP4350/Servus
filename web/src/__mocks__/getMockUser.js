function getMockUser() {
    return {
        username: 'testuser',
        firstName: 'test',
        lastName: 'user',
        password: '123',
        bio: 'im a test user',
    };
}

export const mockUser = getMockUser();
