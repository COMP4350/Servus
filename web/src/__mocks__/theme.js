import { createMuiTheme } from '@material-ui/core/';

export default createMuiTheme({
    palette: {
        primary: {
            main: '#647AA3',
        },
        secondary: {
            main: '#EC5732',
        },
        default: {
            main: '#FFF2EB',
        },
    },
    background: {
        dark: '#151515',
        main: '#272727',
    },
    typographs: {
        fontFamily: ['Roboto'],
    },
});
