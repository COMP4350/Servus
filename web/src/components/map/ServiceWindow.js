import React, { useState, useEffect } from 'react';
import axios from 'axios';
import { makeStyles } from '@material-ui/core/styles';
import { Card, Typography, Button } from '@material-ui/core';
import moment from 'moment';
import MomentUtils from '@date-io/moment';
import { MuiPickersUtilsProvider, DateTimePicker } from '@material-ui/pickers';
import { useCookies } from 'react-cookie';
import { useHistory } from 'react-router-dom';

const useStyles = makeStyles(theme => ({
    container: {
        display: 'flex',
        flexWrap: 'wrap',
    },
    textField: {
        marginLeft: theme.spacing(1),
        marginRight: theme.spacing(1),
        width: 200,
    },
}));

const ServiceWindow = props => {
    const [form, setForm] = useState();
    const [errors, setErrors] = useState({});
    const [cookies] = useCookies();
    const history = useHistory();

    const validate = () => {
        let errors = {};
        if (!form) {
            errors.time = 'time is required';
        }
        setErrors(errors);
        setValid(Object.getOwnPropertyNames(errors).length == 0);
    };

    const classes = useStyles();
    const [valid, setValid] = useState(false);
    const bookAppointment = () => {
        if (valid) {
            axios
                .post(
                    `/appointment/${cookies.username}`,
                    {
                        service_id: props.service._id,
                        provider: props.service.provider,
                        booked_time: form,
                    },
                    {
                        withCredentials: true,
                    }
                )
                .then(res => {
                    if (res.data.errors) {
                        if (res.data.timeslot) {
                            if (res.data.timeslot.availability_start) {
                                alert(
                                    `Please book within timeslot ${res.data.timeslot.availability_start} to ${res.data.timeslot.availability_end}`
                                );
                            } else {
                                alert('Service not available on this day');
                            }
                        } else if (res.data.nocons) {
                            alert(
                                `Conflict with another appointment from time ${moment(
                                    res.data.nocons.conflict_start
                                ).local()} to ${moment(
                                    res.data.nocons.conflict_end
                                ).local()}`
                            );
                        } else {
                            alert(`Error occured when trying to book`);
                        }
                    } else {
                        alert(
                            `Successfully booked Service with ${props.service.provider}`
                        );
                    }
                })
                .catch(err => {
                    throw err;
                });
        }
    };
    useEffect(() => {
        if (cookies.username) bookAppointment();
        else if(history) history.push('/login');
    }, [valid, form]);

    return (
        <Card>
            <Typography color="textSecondary">
                {props.service.provider}
            </Typography>
            <Typography variant="h5">{props.service.name}</Typography>
            <Typography variant="body2" component="p">
                {props.service.description}
            </Typography>
            <form className={classes.container} noValidate>
                <MuiPickersUtilsProvider utils={MomentUtils}>
                    <div className="timepicker">
                        <DateTimePicker
                            label={'Appointment Time'}
                            value={form}
                            initialFocusedDate={Date.now()}
                            ampm={false}
                            onAccept={setForm}
                            onChange={setForm}
                            name="time"
                            autoOk={true}
                            disablePast={true}
                            errors={errors.time}
                        />
                    </div>
                </MuiPickersUtilsProvider>
            </form>
            <Button onClick={validate}>BOOK</Button>
        </Card>
    );
};

export default ServiceWindow;
