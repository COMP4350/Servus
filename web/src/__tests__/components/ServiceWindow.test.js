import React from 'react';
import { act, render } from '@testing-library/react';
import ServiceWindow from '../../components/map/ServiceWindow';
import { mockService } from '../../__mocks__/getMockService';

const blankComponent = () => {
    return <div></div>;
};

jest.mock('../../components/ProfilePicture', () => {
    return {
        __esModule: true,
        default: () => {
            return blankComponent();
        },
    };
});
jest.mock('../../components/map/BookWindow', () => {
    return {
        __esModule: true,
        default: () => {
            return blankComponent();
        },
    };
});

describe('ServiceWindow', () => {
    it('renders correctly', async () => {
        const promise = Promise.resolve();

        const { asFragment } = render(<ServiceWindow service={mockService} />);

        expect(asFragment()).toMatchSnapshot();
        await act(() => promise);
    });
});
