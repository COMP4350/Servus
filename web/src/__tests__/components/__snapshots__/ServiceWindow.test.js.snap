// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceWindow renders correctly 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-window-1"
  >
    <div
      class="makeStyles-upperRow-2"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root makeStyles-iconButton-4"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <div />
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <div
        class="makeStyles-infoBar-6"
      >
        <h1
          class="MuiTypography-root makeStyles-title-7 MuiTypography-h1"
        >
          Risto's Painting
        </h1>
        <p
          class="MuiTypography-root makeStyles-provider-9 MuiTypography-body1 MuiTypography-colorTextSecondary"
        >
          @Risto
        </p>
        <div
          class="MuiBox-root MuiBox-root-14 makeStyles-box-11"
        >
          <span
            class="MuiRating-root"
          >
            <span
              class="MuiRating-decimal"
            >
              <label
                class="MuiRating-label"
                for="customized-color-0-5"
                style="width: 0%; overflow: hidden; z-index: 1; position: absolute;"
              >
                <span
                  class="MuiRating-icon MuiRating-iconFilled WithStyles(ForwardRef(Rating))-iconFilled-15"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiRating-visuallyhidden"
                >
                  0.5 Heart s
                </span>
              </label>
              <input
                class="MuiRating-visuallyhidden"
                id="customized-color-0-5"
                name="customized-color"
                type="radio"
                value="0.5"
              />
              <label
                class="MuiRating-label"
                for="customized-color-1"
              >
                <span
                  class="MuiRating-icon MuiRating-iconFilled WithStyles(ForwardRef(Rating))-iconFilled-15"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiRating-visuallyhidden"
                >
                  1 Heart 
                </span>
              </label>
              <input
                class="MuiRating-visuallyhidden"
                id="customized-color-1"
                name="customized-color"
                type="radio"
                value="1"
              />
            </span>
            <span
              class="MuiRating-decimal"
            >
              <label
                class="MuiRating-label"
                for="customized-color-1-5"
                style="width: 0%; overflow: hidden; z-index: 1; position: absolute;"
              >
                <span
                  class="MuiRating-icon MuiRating-iconFilled WithStyles(ForwardRef(Rating))-iconFilled-15"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiRating-visuallyhidden"
                >
                  1.5 Heart s
                </span>
              </label>
              <input
                class="MuiRating-visuallyhidden"
                id="customized-color-1-5"
                name="customized-color"
                type="radio"
                value="1.5"
              />
              <label
                class="MuiRating-label"
                for="customized-color-2"
              >
                <span
                  class="MuiRating-icon MuiRating-iconFilled WithStyles(ForwardRef(Rating))-iconFilled-15"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiRating-visuallyhidden"
                >
                  2 Heart s
                </span>
              </label>
              <input
                class="MuiRating-visuallyhidden"
                id="customized-color-2"
                name="customized-color"
                type="radio"
                value="2"
              />
            </span>
            <span
              class="MuiRating-decimal"
            >
              <label
                class="MuiRating-label"
                for="customized-color-2-5"
                style="width: 0%; overflow: hidden; z-index: 1; position: absolute;"
              >
                <span
                  class="MuiRating-icon MuiRating-iconFilled WithStyles(ForwardRef(Rating))-iconFilled-15"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiRating-visuallyhidden"
                >
                  2.5 Heart s
                </span>
              </label>
              <input
                class="MuiRating-visuallyhidden"
                id="customized-color-2-5"
                name="customized-color"
                type="radio"
                value="2.5"
              />
              <label
                class="MuiRating-label"
                for="customized-color-3"
              >
                <span
                  class="MuiRating-icon MuiRating-iconFilled WithStyles(ForwardRef(Rating))-iconFilled-15"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiRating-visuallyhidden"
                >
                  3 Heart s
                </span>
              </label>
              <input
                class="MuiRating-visuallyhidden"
                id="customized-color-3"
                name="customized-color"
                type="radio"
                value="3"
              />
            </span>
            <span
              class="MuiRating-decimal"
            >
              <label
                class="MuiRating-label"
                for="customized-color-3-5"
                style="width: 0%; overflow: hidden; z-index: 1; position: absolute;"
              >
                <span
                  class="MuiRating-icon MuiRating-iconFilled WithStyles(ForwardRef(Rating))-iconFilled-15"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiRating-visuallyhidden"
                >
                  3.5 Heart s
                </span>
              </label>
              <input
                class="MuiRating-visuallyhidden"
                id="customized-color-3-5"
                name="customized-color"
                type="radio"
                value="3.5"
              />
              <label
                class="MuiRating-label"
                for="customized-color-4"
              >
                <span
                  class="MuiRating-icon MuiRating-iconFilled WithStyles(ForwardRef(Rating))-iconFilled-15"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiRating-visuallyhidden"
                >
                  4 Heart s
                </span>
              </label>
              <input
                class="MuiRating-visuallyhidden"
                id="customized-color-4"
                name="customized-color"
                type="radio"
                value="4"
              />
            </span>
            <span
              class="MuiRating-decimal"
            >
              <label
                class="MuiRating-label"
                for="customized-color-4-5"
                style="width: 0%; overflow: hidden; z-index: 1; position: absolute;"
              >
                <span
                  class="MuiRating-icon MuiRating-iconFilled WithStyles(ForwardRef(Rating))-iconFilled-15"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiRating-visuallyhidden"
                >
                  4.5 Heart s
                </span>
              </label>
              <input
                class="MuiRating-visuallyhidden"
                id="customized-color-4-5"
                name="customized-color"
                type="radio"
                value="4.5"
              />
              <label
                class="MuiRating-label"
                for="customized-color-5"
              >
                <span
                  class="MuiRating-icon MuiRating-iconFilled WithStyles(ForwardRef(Rating))-iconFilled-15"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiRating-visuallyhidden"
                >
                  5 Heart s
                </span>
              </label>
              <input
                checked=""
                class="MuiRating-visuallyhidden"
                id="customized-color-5"
                name="customized-color"
                type="radio"
                value="5"
              />
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      align="left"
      class="makeStyles-lowerRow-3"
    >
      <div
        class="makeStyles-info-5"
      >
        <p
          class="MuiTypography-root makeStyles-desc-10 MuiTypography-body2"
        >
          Cost: 100
          <br />
          Risto paints you
        </p>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-book-13 MuiButton-containedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Book
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
